apis/s3/src/main/java/org/jclouds/s3/filters/RequestAuthorizeSignatureV4.java

```diff
@@ -0,0 +1,574 @@
+
+        Credentials credentials = creds.get();
+
+        if (credentials instanceof SessionCredentials) {
+            String token = SessionCredentials.class.cast(credentials).getSessionToken();
+            requestBuilder.replaceHeader("X-Amz-Security-Token", token);
+            signedHeadersBuilder.put("x-amz-security-token", token);
+        }
+
+        String contentSha256 = calculateContentHash(requestBuilder, request.getMethod(), endpoint, payload);
+        if ("required".equals(requestBuilder.build().getFirstHeaderOrNull("x-amz-content-sha256"))) {
+            requestBuilder.replaceHeader("x-amz-content-sha256", contentSha256);
+            signedHeadersBuilder.put("x-amz-content-sha256", contentSha256);
+        }
+
+        // put x-amz-date
+        signedHeadersBuilder.put("x-amz-date", timestamp);
+
+
+        requestBuilder.payload(payload);
+
+        ImmutableMap<String, String> signedHeaders = signedHeadersBuilder.build();
+
+        String stringToSign = createStringToSign(requestBuilder.build(), signedHeaders, credentialScope);
+        signatureWire.getWireLog().debug("<< " + stringToSign);
+
+        byte[] signatureKey = signatureKey(credentials.credential, datestamp, region, service);
+        String signature = base16().lowerCase().encode(hmacSHA256(stringToSign, signatureKey));
+
+        StringBuilder authorization = new StringBuilder("AWS4-HMAC-SHA256 ");
+        authorization.append("Credential=").append(credentials.identity).append('/').append(credentialScope).append(", ");
+        authorization.append("SignedHeaders=").append(Joiner.on(';').join(signedHeaders.keySet())).append(", ");
+        authorization.append("Signature=").append(signature);
+
+        return requestBuilder.addHeader(AUTHORIZATION, authorization.toString()).build();
+    }
+
+    // append all of 'x-amz-*' headers
+    void appendAmzHeaders(HttpRequest request, ImmutableMap.Builder<String, String> signedHeadersBuilder) {
+        for (Map.Entry<String, String> header : request.getHeaders().entries()) {
+            String key = header.getKey();
+            if (key.startsWith("x-" + headerTag + "-")) {
+                signedHeadersBuilder.put(key.toLowerCase(), header.getValue());
+            }
+        }
+    }
+
+    // calculate content hash
+    protected String calculateContentHash(
+        HttpRequest.Builder requestBuilder,
+        String method,
+        URI endpoint,
+        Payload payload
+    ) {
+        InputStream payloadStream;
+        try {
+            payloadStream = usePayloadForQueryParameters(method, payload) ?
+                getQueryStringContent(endpoint)
+                : getPayloadContentWithoutQueryString(payload);
+        } catch (IOException e) {
+            throw new HttpException("Unable to open stream before calculate AWS4 signature", e);
+        }
+        String contentSha256 = base16().lowerCase().encode(hash(payloadStream));
+        try {
+            payloadStream.reset();
```

Comment:
payload stream use calculate content hash.
if can not be repeatable, the payload cannot append to http request body.